packages/server/lib/cloud/upload/StreamActivityMonitor.ts

```diff
@@ -0,0 +1,75 @@
+import Debug from 'debug'
+import { Transform, Readable } from 'stream'
+
+const debug = Debug('cypress:server:cloud:stream-activity-monitor')
+const debugVerbose = Debug('cypress-verbose:server:cloud:stream-activity-monitor')
+
+export class StreamStartTimedOutError extends Error {
+  constructor (maxStartDwellTime: number) {
+    super(`Source stream failed to begin sending data after ${maxStartDwellTime}ms`)
+  }
+}
+
+export class StreamStalledError extends Error {
+  constructor (maxActivityDwellTime: number) {
+    super(`Stream stalled: no activity detected in the previous ${maxActivityDwellTime}ms`)
+  }
+}
+
+export class StreamActivityMonitor {
+  private streamMonitor: Transform | undefined
+  private startTimeout: NodeJS.Timeout | undefined
+  private activityTimeout: NodeJS.Timeout | undefined
+  private controller: AbortController
+
+  constructor (private maxStartDwellTime: number, private maxActivityDwellTime: number) {
+    this.controller = new AbortController()
+  }
+
+  public getController () {
+    return this.controller
+  }
+
+  public monitor (stream: Readable): Readable {
+    debug('monitoring stream')
+    if (this.streamMonitor || this.startTimeout || this.activityTimeout) {
+      this.reset()
+    }
+
+    this.streamMonitor = new Transform({
```

Comment:
I'll be honest... I'm somewhat illiterate when it comes to readable streams. What exactly is going on here?

Reply:
this in the description is helpful
> implements a StreamActivityMonitor that produces a Transform stream that monitors a given Readable, and an AbortController that will be signalled if the stream is inactive for too long. This is used in uploadStream.ts to detect stalled uploads, by piping the Test Replay archive's fs.ReadStream through the StreamActivityMonitor's Transform.

Are we able to add some comments here to describe what is happening?

Reply:
I'll definitely add some comments! Streams are wacky to wrap your head around, especially since there are two standards (node streams and whatwg streams)

Reply:
4cc05bd7f3cce36ad85c8a7fd51675edcc19891a is incredibly helpful. thank you!